apps/blockchain/lib/blockchain/ethash.ex

```diff
@@ -80,10 +85,25 @@ defmodule Blockchain.Ethash do
     |> is_nil
   end
 
+  @doc """
+  Calculates the cache, c, outlined in Appendix J section J.3.2 of the Yellow
+  Paper, by performing the RandMemoHash algorithm n times (where n is 3 if it is
+  j_cacherounds)
+  """
+  @spec calculate_cache(cache(), integer()) :: cache()
+  def calculate_cache(initial_cache, number_of_rounds \\ @j_cacherounds)
```

Comment:
whats the reason this is public?

Reply:
Probably to allow unit testing. So is life with these things. Though you could move the function to a sub module to reduce exposure, it’s probably fine to have public.

Reply:
Yeah, moving it to a separate module or testing the consequence is preferred I think.

Reply:
I would argue, overall, that public/private functions aren't generally worth a refactor unless you expect people to be using your API and it would get in the way. Since this module itself is effectively private, I would argue we should keep it as it is unless it gets out of hand.

Reply:
If the module is effectively private we can make sure that functions stay private too, I think. 
Static code analysis is also unable to report whether they’re being used or not which affects refactoring and maintainability. 
Chaning private functions also has a bigger side effect on tests, which shouldn’t be the case.

Reply:
We should move this conversation off-line, but I tend to say let the original coder decide such issues.

Reply:
@InoMurko I am glad you're bringing up the point of public functions that shouldn't be! I totally agree with you. My approach right now is to leave some of these public _in the meantime_. 

I'm trying to implement the Ethash algorithm piece by piece so it's easier for others to review (and it is easier for me to understand too), but once the algorithm is fully implemented, I do intend to make all functions that should be private into private functions, and I'll remove the appropriate tests. 

As we go along, if it makes sense to extract some functionality into other modules, I'm happy to do that. I'm always on board of refactoring as we go. 

Does that work for you?

Reply:
Great! 👍🏻